validate.go

```diff
@@ -18,34 +19,127 @@ func validate(payload []byte) ([]byte, error) {
+
+			if mount.Exists() {
+				// is a hostpath and is in use, validate against settings
+
+				readOnly := false // volumeMount.readOnly missing means 'false'
+				if gjson.Get(mount.String(), "readOnly").Exists() {
+					readOnly = gjson.Get(mount.String(), "readOnly").Bool()
+				}
+
+				path := gjson.Get(volume.String(), "hostPath.path").String()
+
+				match := false
+				// readOnly attribute of most specific AllowedHostPath takes precendence:
+				previousAllowedHostPath := ""
+				for _, allowedHostPath := range settings.AllowedHostPaths {
+					if hasPathPrefix(path, allowedHostPath.PathPrefix) {
+						// current setting allowedHostPath matches path of volumeMount
+						if hasPathPrefix(allowedHostPath.PathPrefix, previousAllowedHostPath) {
+							// allowedHostPath is more specific (and has precendence over
+							//	past allowedHostPath), or the same path
+							match = true
+							err = validatePath(path, readOnly, allowedHostPath)
+							previousAllowedHostPath = allowedHostPath.PathPrefix
+						}
+					}
+				}
+				if !match {
+					// path didn't match against any PathPrefix in settings
+					err = fmt.Errorf("hostPath '%s' is not in the AllowedHostPaths list",
+						path)
+				}
+				if err != nil {
+					break // stop first for, a path was not allowed against all settings, reject
+				}
+			}
+		}
+	}
+
+	if err != nil {
+		logger.DebugWithFields("rejecting pod object", func(e onelog.Entry) {
+			name := gjson.GetBytes(payload, "request.object.metadata.name").String()
+			namespace := gjson.
+				GetBytes(payload, "request.object.metadata.namespace").String()
 			e.String("name", name)
-			e.String("denied_names", settings.DeniedNames.String())
+			e.String("namespace", namespace)
 		})
-
 		return kubewarden.RejectRequest(
-			kubewarden.Message(
-				fmt.Sprintf("The '%s' name is on the deny list", name)),
+			kubewarden.Message(err.Error()),
 			kubewarden.NoCode)
 	}
 
-	logger.Info("accepting ingress object")
 	return kubewarden.AcceptRequest()
 }
+
+// validatePath validates the path prefix and its readOnly state against the
+// passed hostPath, and returns a matching error if failed.
+func validatePath(path string, readOnly bool, hostPath HostPath) (err error) {
+	if hasPathPrefix(path, hostPath.PathPrefix) {
+		if readOnly != hostPath.ReadOnly {
+			return fmt.Errorf("hostPath '%s' should be readOnly '%t'",
+				path, hostPath.ReadOnly)
+		}
+	}
+	return nil
+}
+
+func hasPathPrefix(path string, prefix string) bool {
+	// allow "/foo", "/foo/", "/foo/bar", etc
+	// disallow "/fool", "/etc/foo", etc
+	// "/foo/../" is never valid.
+	// Hence, ensure paths terminate in `/`:
+	pathTerminated := path
+	if !strings.HasSuffix(pathTerminated, "/") {
+		pathTerminated = pathTerminated + "/"
+	}
+	prefixTerminated := prefix
+	if !strings.HasSuffix(prefixTerminated, "/") {
+		prefixTerminated = prefixTerminated + "/"
+	}
+	return strings.HasPrefix(pathTerminated, prefixTerminated)
+}
```

Comment:
Can we use something from the golang stdlib that does this?

Reply:
There was `filepath.HasPrefix(p, prefix string)`, but got [deprecated](https://pkg.go.dev/path/filepath#HasPrefix). And the implementation is even worse: https://github.com/golang/go/blob/1f9c9d853067635305f72e247c5b49e3fa5da8af/src/path/filepath/path_unix.go#L27-L29

Reply:
Yeah, I did a quick research when doing my review, but I didn't find anything better. I think it's fine to use the current code